### PR TITLE
Cans 183 edit child form

### PR DESCRIPTION
**Pull request by abelayt**
## Tests

- [x] I used TDD to develop the feature
- [x ] I have included unit tests
- [ ] I have included new acceptance tests or modified existing ones
- [ ] I have included other tests

## Types of changes

- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactoring (No behavioral changes)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [x] I ran all my tests locally which were green.
- [ ] My code adds no new issues to Code Climate
- [x\] I ran all the linters for the project.
- [x] My code is in a stable state ready to be deployable, but not necessarily complete.
- [x] I promise on my honor as a CWDS developer to ensure I don't break the build (or bring donuts if I leave things broken), to check linters, use best practices, to leave the code in better shape than I found it, and to have a good day.

### CODE REVIEW

File: app/javascript/Application/components/Client/Client.js [hunk 1]
```diff
@@ -67,16 +82,50 @@ class Client extends Component {
   };
 
   render() {
-    const childData = this.state.childData;
+    // const childData = this.state.childData;
+    // const { childData, shouldRenderClientEditMessage, shouldRenderClientAddMessage } = this.state;
```
Comment (jonbwalker) [inline app/javascript/Application/components/Client/Client.js]: remove commented out code

File: app/javascript/Application/components/Client/ClientEditForm.js [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+                    <MenuItem key={option.id} value={option} className={classes.menu}>
+                      <span id={'county-name'}>{option.name}</span>
+                    </MenuItem>
+                  ))}
+                </TextField>
+              </form>
+            )}
+          </CardContent>
+          <CardActions>
+            <a onClick={this.handleCancel} href={`/clients/${childInfo.id}`}>
```
Comment (jonbwalker) [inline app/javascript/Application/components/Client/ClientEditForm.js]: We should use the react-router Link component here

File: app/javascript/Application/components/Client/Client.js [hunk 3]
```diff
@@ -67,16 +82,50 @@ class Client extends Component {
...
-              <CardHeader className={'card-header-cans'} title="Child/Youth Information" />
+              <CardHeader
+                className={'card-header-cans'}
+                title="Child/Youth Information"
+                action={
+                  <Button
+                    size="small"
+                    color="inherit"
+                    className={'card-header-cans-button'}
+                    href={`/clients/edit/${childData.id}`}
```
Comment (jonbwalker) [inline app/javascript/Application/components/Client/Client.js]: We should leverage react-router Link by wrapping the Button with it.

File: app/javascript/Application/components/Client/ClientAddForm.js [hunk 4]
```diff
@@ -276,24 +287,31 @@ class ChildForm extends Component {
...
-        )}
       </Fragment>
     );
   }
 }
 
-ChildForm.propTypes = {
+ClientAddForm.propTypes = {
   classes: PropTypes.object.isRequired,
+  match: PropTypes.object.isRequired,
```
Comment (jonbwalker) [inline app/javascript/Application/components/Client/ClientAddForm.js]: If match is required, then we shouldn't have default props for it.

File: app/javascript/Application/components/Client/ClientEditForm.js [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+          </CardActions>
+        </Card>
+      </Fragment>
+    );
+  }
+}
+
+ClientEditForm.propTypes = {
+  classes: PropTypes.object.isRequired,
+  match: PropTypes.object.isRequired,
```
Comment (jonbwalker) [inline app/javascript/Application/components/Client/ClientEditForm.js]: same idea with required props

File: app/javascript/Application/components/Client/ClientEditForm.test.js [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+  });
+
+  it('#handleSubmit()', async () => {
+    const ClientServicePutSpy = jest.spyOn(ClientService, 'updateClient');
+    const wrapper = getWrapper();
+    const ClientEditWrapper = wrapper.find('ClientEditForm').dive();
+    ClientServicePutSpy.mockReturnValue(Promise.resolve(childInfoJson));
+    await ClientEditWrapper.instance().handleSubmit();
+    expect(ClientEditWrapper.state().redirection.shouldRedirect).toBe(true);
+  });
```
Comment (jonbwalker) [inline app/javascript/Application/components/Client/ClientEditForm.test.js]: The spy needs to be cleared after the test is finished. 
ClientServicePutSpy.mockClear();

File: app/javascript/Application/components/Layout/SideNav.js [hunk 7]
```diff
@@ -7,8 +7,8 @@ const URL_MAP = [
   { url: '/', text: 'Child & Family Teams' },
   { url: '/reports', text: 'Reports' },
   { url: '/clients/new', text: 'Add Child' },
+  // { url: '/clients/edit/:id', text: 'Edit Child' },
```
Comment (jonbwalker) [inline app/javascript/Application/components/Layout/SideNav.js]: remove commented out code

File: docker-compose.yml [hunk 8]
```diff
@@ -19,12 +19,10 @@ services:
       PERRY_PORT: ${PERRY_PORT}
```
Comment (jonbwalker) [inline docker-compose.yml]: I'm assuming this wasn't meant to be committed.

File: app/javascript/Application/components/Client/ClientAddForm.js (hunk 4 above)
Comment (abelayt) [reply app/javascript/Application/components/Client/ClientAddForm.js]: done

File: app/javascript/Application/components/Layout/SideNav.js (hunk 7 above)
Comment (abelayt) [reply app/javascript/Application/components/Layout/SideNav.js]: done

File: app/javascript/Application/components/Client/ClientEditForm.test.js (hunk 6 above)
Comment (abelayt) [reply app/javascript/Application/components/Client/ClientEditForm.test.js]: done, thanks!

File: app/javascript/Application/components/Client/ClientEditForm.js (hunk 5 above)
Comment (abelayt) [reply app/javascript/Application/components/Client/ClientEditForm.js]: done

File: app/javascript/Application/components/Client/Client.js (hunk 1 above)
Comment (abelayt) [reply app/javascript/Application/components/Client/Client.js]: done

File: app/javascript/Application/components/Client/ClientEditForm.js (hunk 2 above)
Comment (abelayt) [reply app/javascript/Application/components/Client/ClientEditForm.js]: done

File: app/javascript/Application/components/Client/Client.js (hunk 3 above)
Comment (abelayt) [reply app/javascript/Application/components/Client/Client.js]: done

File: app/javascript/Application/components/Client/Client.js [hunk 9]
```diff
@@ -67,16 +83,44 @@ class Client extends Component {
   };
 
   render() {
-    const childData = this.state.childData;
+    const { childData, shouldRenderClientAddMessage, shouldRenderClientEditMessage } = this.state;
```
Comment (edgibbs) [inline app/javascript/Application/components/Client/Client.js]: I like these variable names, very clear as to their purpose.  :sunny:

File: app/javascript/Application/components/Client/ClientAddForm.js [hunk 10]
```diff
@@ -158,11 +170,13 @@ class ChildForm extends Component {
...
-    if (this.state.navigate) {
-      return <Redirect to={`/clients/${this.state.childInfo.id}`} />;
-    }
-
     const { classes } = this.props;
+    const { childInfo, childInfoValidation, counties, isSaveButtonDisabled, redirection } = this.state;
+    const { shouldRedirect, successClientAddId } = redirection;
+
+    if (shouldRedirect) {
+      return <Redirect push to={{ pathname: `/clients/${childInfo.id}`, state: { successClientAddId } }} />;
```
Comment (edgibbs) [inline app/javascript/Application/components/Client/ClientAddForm.js]: This check needs a test, I commented it out and nothing failed.

File: app/javascript/Application/components/Client/ClientEditForm.js (hunk 5 above)
Comment (edgibbs) [reply app/javascript/Application/components/Client/ClientEditForm.js]: So not sure this is in scope to change for this PR, but this component is much like ClientAddForm, which is why code climate went a bit crazy on dup violations.  I think a general component for the form part makes more sense or a single component that has to manage a bit of state.

File: lib/people/people_repository.rb [hunk 11]
```diff
@@ -15,6 +15,10 @@ def create(payload)
       @http_service.call('/people', :post, @token, payload)
     end
 
+    def update(id, payload)
```
Comment (edgibbs) [inline lib/people/people_repository.rb]: This needs a test.

File: docker-compose.yml [hunk 12]
```diff
@@ -71,4 +71,4 @@ services:
         source: ./spec
         target: /app/spec
     environment:
-      CANS_WEB_BASE_URL: ${CANS_WEB_BASE_URL}
+      CANS_WEB_BASE_URL: ${CANS_WEB_BASE_URL}
```
Comment (edgibbs) [inline docker-compose.yml]: newline issue

File: app/controllers/api/people_controller.rb [hunk 13]
```diff
@@ -17,5 +17,11 @@ def create
       response = people_repository.create(params[:person])
       render json: response.body, status: response.status
     end
+
+    def update
+      people_repository = People::PeopleRepository.new(session[:token])
```
[the remark this answers is not in the document]
Comment (abelayt) [reply app/controllers/api/people_controller.rb]: done

**Comment by edgibbs** [comment]
Approving changes for now will look at refactor when we swap out CWDS component cards to reduce duplication.

File: lib/people/people_repository.rb (hunk 11 above)
Comment (abelayt) [reply lib/people/people_repository.rb]: done

**Pull request merged**